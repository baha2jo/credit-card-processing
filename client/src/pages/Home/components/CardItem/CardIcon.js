import React from 'react';
import { string } from 'prop-types';
import { getCardType } from './card.helper';

import { ReactComponent as Visa } from 'assets/svgs/visa.svg';
import { ReactComponent as Master } from 'assets/svgs/master.svg';
import { ReactComponent as Discover } from 'assets/svgs/discover.svg';
import { ReactComponent as Amex } from 'assets/svgs/amex.svg';
import { ReactComponent as DefaultCard } from 'assets/svgs/defaultcard.svg';

const CardIcon = ({ number }) => {
  const cardType = getCardType(number);

  const renderCardIcon = () => {
    switch (cardType) {
      case 'Visa':
        return <Visa />;
      case 'Mastercard':
        return <Master />;
      case 'AMEX':
        return <Amex />;
      case 'Discover':
        return <Discover />;
      default:
        return <DefaultCard />;
    }
  };

  return <>{renderCardIcon()}</>;
};

CardIcon.propTypes = {
  number: string.isRequired
};

export default CardIcon;
