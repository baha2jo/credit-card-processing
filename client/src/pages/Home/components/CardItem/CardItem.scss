.card-item {
  &__icon {
    height: 100px;
    width: 100px;
    margin-right: 24px;
  }
}
