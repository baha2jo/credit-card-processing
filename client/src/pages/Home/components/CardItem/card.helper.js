// inspired from https://stackoverflow.com/questions/5911236/identify-card-type-from-card-number

export const getCardType = number => {
  // visa
  var re = new RegExp('^4');
  if (number.match(re) != null) return 'Visa';

  // Mastercard
  // Updated for Mastercard 2017 BINs expansion
  if (
    /^(5[1-5][0-9]{14}|2(22[1-9][0-9]{12}|2[3-9][0-9]{13}|[3-6][0-9]{14}|7[0-1][0-9]{13}|720[0-9]{12}))$/.test(
      number
    )
  )
    return 'Mastercard';

  // AMEX
  re = new RegExp('^3[47]');
  if (number.match(re) != null) return 'AMEX';

  // Discover
  re = new RegExp(
    '^(6011|622(12[6-9]|1[3-9][0-9]|[2-8][0-9]{2}|9[0-1][0-9]|92[0-5]|64[4-9])|65)'
  );
  if (number.match(re) != null) return 'Discover';

  // Diners
  re = new RegExp('^36');
  if (number.match(re) != null) return 'Diners';

  // Diners - Carte Blanche
  re = new RegExp('^30[0-5]');
  if (number.match(re) != null) return 'Diners - Carte Blanche';

  return '';
};
