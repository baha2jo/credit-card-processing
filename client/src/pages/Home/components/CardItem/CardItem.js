import React from 'react';
import CardIcon from './CardIcon';
import './CardItem.scss';
import { shape, string } from 'prop-types';

export const CardItem = ({ card: { name, cardNumber, limit, expiryDate } }) => {
  return (
    <div className='card-item'>
      <div className='d-flex align-items-center'>
        <span className='card-item__icon mr-4'>
          <CardIcon number={cardNumber} />
        </span>
        <div>
          <p className='mb-0 card-number'>{cardNumber}</p>
          <p className='mb-0 card-holder'>{name}</p>
          <small className='text-muted card-expiry'>{expiryDate}</small>
        </div>
      </div>
      <p className='text-right text-info card-limit'>{limit}$</p>

      <div className='line'></div>
    </div>
  );
};

CardItem.propTypes = {
  card: shape({
    name: string,
    cardNumber: string,
    limit: string,
    expiryDate: string
  }).isRequired
};
