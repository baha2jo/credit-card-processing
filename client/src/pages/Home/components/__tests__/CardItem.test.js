import { CardItem } from '../CardItem';

const card = {
  cardNumber: '4970856264139166',
  expiryDate: '12/21',
  limit: '122',
  name: '3213',
  _id: '2QCfp3SjZ7bMo1gi'
};

function setup() {
  const component = shallow(<CardItem card={card} />);

  return component;
}

describe('CardItem component', () => {
  const component = setup();

  it('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should display the correct card number', () => {
    expect(component.find('.card-number').text()).toBe(card.cardNumber);
  });

  it('should display the correct card holder name', () => {
    expect(component.find('.card-holder').text()).toBe(card.name);
  });

  it('should display the correct expiry date', () => {
    expect(component.find('.card-expiry').text()).toBe(card.expiryDate);
  });

  it('should display the correct card limit', () => {
    expect(component.find('.card-limit').text()).toBe(`${card.limit}$`);
  });
});
