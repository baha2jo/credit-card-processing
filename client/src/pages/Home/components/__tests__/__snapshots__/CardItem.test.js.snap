// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem component should render correctly 1`] = `
<div
  className="card-item"
>
  <div
    className="d-flex align-items-center"
  >
    <span
      className="card-item__icon mr-4"
    >
      <CardIcon
        number="4970856264139166"
      />
    </span>
    <div>
      <p
        className="mb-0 card-number"
      >
        4970856264139166
      </p>
      <p
        className="mb-0 card-holder"
      >
        3213
      </p>
      <small
        className="text-muted card-expiry"
      >
        12/21
      </small>
    </div>
  </div>
  <p
    className="text-right text-info card-limit"
  >
    122
    $
  </p>
  <div
    className="line"
  />
</div>
`;
